source/dem/plane_insertion.cc

```diff
@@ -0,0 +1,281 @@
+      // this time step. This is basically the only information share between
+      // processor.
+      auto number_of_particles_to_insert_per_core =
+        Utilities::MPI::all_gather(communicator, empty_cells_on_proc.size());
+
+      // Every processor does the following calculation. This could be done by a
+      // "gather and scatter", but it's probably less efficient. First, we loop
+      // over the vector that we have created using the all_gather. Every
+      // integer in that vector gives the number of empty_cell of a certain
+      // processor.
+      for (unsigned int i = 0;
+           i < number_of_particles_to_insert_per_core.size();
+           ++i)
+        {
+          // We subtract those integer.
+          remained_particles_of_each_type -=
+            number_of_particles_to_insert_per_core[i];
+
+          // If the remained_particles_of_each_type become negative, this means
+          // that we are trying to insert to many particle in the simulation.
+          // Thus, we need to decrease the number of empty_cell at this time
+          // step. So...
+          if (remained_particles_of_each_type <= 0)
+            {
+              // ... we decrease the number of particle at the current
+              // processor, since it's the one who got the variable to become
+              // negative. Then...
+              number_of_particles_to_insert_per_core[i] +=
+                remained_particles_of_each_type;
+
+              // ... we put remained_particles_of_each_type back to zero.
+              remained_particles_of_each_type = 0;
+
+              // Loop over the remaining processor and put those at zero.
+              for (unsigned int j = i + 1;
+                   j < number_of_particles_to_insert_per_core.size();
+                   ++j)
+                {
+                  number_of_particles_to_insert_per_core[j] = 0;
+                }
+              break;
+            }
+        }
+
+      // Now, every processor knows how many particle it needs to insert. They
+      // also know how many particle the other processor have to insert. This
+      // way, it's possible for a processor to find the correct range of
+      // particle id to use with the particle_counter variable.
+
+      // This is the starting id of the processor.
+      unsigned int id_on_proc =
+        std::accumulate(number_of_particles_to_insert_per_core.begin(),
+                        number_of_particles_to_insert_per_core.begin() +
+                          this_mpi_process,
+                        particle_counter);
+
+      // We yet didn't say which empty cells won't be use anymore.
+      // empty_cells_on_proc needs to be shorter for some processor.
+      while (empty_cells_on_proc.size() <
+             number_of_particles_to_insert_per_core[this_mpi_process])
+        {
+          auto it = empty_cells_on_proc.end();
+          empty_cells_on_proc.erase(it);
+        }
```

Comment:
Here I think I would juste extract part of the vector like or remove in 1 line with something like:
empty_cells_on_proc.erase(empty_cells_on_proc.begin() + number_of_particles_to_insert_per_core[this_mpi_process], empty_cells_on_proc.end())